crypto/BRCryptoPayment.c

```diff
@@ -452,59 +477,144 @@ cryptoPaymentProtocolRequestGetPrimaryTargetAddress (BRCryptoPaymentProtocolRequ
+            array_new (certs, 10);
+            array_new (certLens, 10);
+
+            size_t certLen, index = 0;
+            while (( certLen = BRPaymentProtocolRequestCert (protoReq->u.btc.request, NULL, 0, index)) > 0) {
+                uint8_t *cert = malloc (certLen);
+                BRPaymentProtocolRequestCert (protoReq->u.btc.request, cert, certLen, index);
+
+                array_add (certs, cert);
+                array_add (certLens, certLen);
+
+                index++;
+            }
+
+            error = protoReq->u.btc.callbacks.validator (protoReq,
+                                                         protoReq->u.btc.callbacks.context,
+                                                         protoReq->u.btc.request->pkiType,
+                                                         protoReq->u.btc.request->details->expires,
+                                                         certs, certLens, array_count (certs),
+                                                         digest, digestLen,
+                                                         protoReq->u.btc.request->signature, protoReq->u.btc.request->sigLen);
+
+            for (index = 0; index < array_count(certs); index++) {
+                free (certs[index]);
+            }
+            array_free (certs);
+            array_free (certLens);
+
+            if (digest) {
+                free (digest);
+            }
+            break;
+        }
+        default: {
+            assert (0);
+            break;
+        }
+    }
+    return error;
 }
 
 private_extern BRArrayOf(BRTxOutput)
 cryptoPaymentProtocolRequestGetOutputsAsBTC (BRCryptoPaymentProtocolRequest protoReq) {
+    BRArrayOf(BRTxOutput) outputs = NULL;
     switch (protoReq->type) {
         case CRYPTO_PAYMENT_PROTOCOL_TYPE_BITPAY:
         case CRYPTO_PAYMENT_PROTOCOL_TYPE_BIP70: {
             BRPaymentProtocolRequest *request = protoReq->u.btc.request;
-            BRArrayOf(BRTxOutput) outputs;
-            array_new(outputs, request->details->outCount);
-            array_add_array(outputs, request->details->outputs, request->details->outCount);
-            return outputs;
+            if (0 != request->details->outCount) {
+                array_new(outputs, request->details->outCount);
+                array_add_array(outputs, request->details->outputs, request->details->outCount);
+            }
+            break;
         }
         default: {
             assert (0);
             break;
         }
     }
-    return NULL;
+    return outputs;
 }
 
 static uint8_t*
 cryptoPaymentProtocolRequestGetMerchantData (BRCryptoPaymentProtocolRequest protoReq, size_t *merchantDataLen) {
+    uint8_t *merchantData = NULL;
     switch (protoReq->type) {
         case CRYPTO_PAYMENT_PROTOCOL_TYPE_BITPAY:
         case CRYPTO_PAYMENT_PROTOCOL_TYPE_BIP70: {
             BRPaymentProtocolRequest *request = protoReq->u.btc.request;
+            merchantData = request->details->merchantData;
```

Comment:
The return type does not include 'const' but `details->merchantData` is not copied?  At the call site, it looks like 'const' would work for this function.

Reply:
Good catch. Will fix.